components/brave_federated/trace_collection_service.cc

```diff
@@ -0,0 +1,205 @@
+        semantics {
+          sender: "User Trace Collection Service"
+          description:
+            "Report of anonymized usage statistics. For more info see "
+            "TODO: https://wikilink_here"
+          trigger:
+            "Reports are automatically generated on startup and at intervals "
+            "while Brave is running."
+          data:
+            "A protocol buffer with anonymized and encrypted usage data."
+          destination: WEBSITE
+        }
+        policy {
+          cookies_allowed: NO
+          setting:
+            "Users can enable or disable it in brave://settings/privacy"
+          policy_exception_justification:
+            "Not implemented."
+        }
+    )");
+}
+
+} // anonymous namespace
+
+BraveTraceCollectionService::BraveTraceCollectionService(PrefService* pref_service, 
+                            scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory)
+        : pref_service_(pref_service), url_loader_factory_(url_loader_factory) {}
+
+BraveTraceCollectionService::~BraveTraceCollectionService() {}
+
+void BraveTraceCollectionService::RegisterLocalStatePrefs(PrefRegistrySimple* registry) {
+  registry->RegisterIntegerPref(kLastCheckedSlot, -1);
+  registry->RegisterStringPref(kEphemeralID, {});
+}
+
+void BraveTraceCollectionService::Start() {
+    if (!isAdsEnabled() || !isP3AEnabled() || !isTraceCollectionEnabled()) {
+        return;
+    }
+
+    std::cerr << "FL-Trace-Collection: is active! \n";
+    LoadPrefs();
+
+    // Fake update timer for current slot
+    DCHECK(!fake_update_activity_timer_);
+    fake_update_activity_timer_ = std::make_unique<base::RetainingOneShotTimer>();
+    fake_update_activity_timer_->Start(
+        FROM_HERE,
+        base::TimeDelta::FromSeconds(fake_update_duration_*60),
+        this, &BraveTraceCollectionService::OnFakeUpdateTimerFired
+    );
+    
+    // Recurrent timer
+    DCHECK(!collection_slot_periodic_timer_);
+    collection_slot_periodic_timer_ = std::make_unique<base::RepeatingTimer>();
+    collection_slot_periodic_timer_->Start(
+        FROM_HERE,
+        base::TimeDelta::FromSeconds(collection_slot_size_*60/2),
+        this, &BraveTraceCollectionService::OnCollectionSlotStartTimerFired
+    );
+}
+
+void BraveTraceCollectionService::LoadPrefs() {
+
+    collection_slot_size_ = trace_collection::features::GetCollectionSlotSizeValue();
+    fake_update_duration_ = trace_collection::features::GetFakeUpdateDurationValue();
+    trace_collection_endpoint_ = GURL(federatedLearningUrl);
+
+    last_checked_slot_ = pref_service_->GetInteger(kLastCheckedSlot);
+    platform_ = GetPlatformIdentifier();
+    std::string ephemeral_ID = pref_service_->GetString(kEphemeralID);
+    if (ephemeral_ID.empty()) {
+        ephemeral_ID_ = 
+                base::ToUpperASCII(base::UnguessableToken::Create().ToString());
+        // TODO(lminto): May want to set expiration date for ephemeral ID, after which
```

Comment:
I think we should either implement this or implement an automatic kill-switch (independent of griffin) to stop collecting after the allowed period

Reply:
we could allow the expiration to be set via Griffin and set a default expiration on the client, similarly to what I did for `collection_slot_size`. Wdyt?

Reply:
I like it, let's do that @LorenzoMinto